coordinator/src/main/scala/filodb.coordinator/NodeCoordinatorActor.scala

```diff
@@ -0,0 +1,229 @@
+    if (datasetObj.projections.isEmpty) {
+      originator ! DatasetError(s"There must be at least one projection in dataset $datasetObj")
+    } else {
+      (for { resp1 <- metaStore.newDataset(datasetObj)
+             resp2 <- Future.sequence(columns.map(metaStore.newColumn(_)))
+             resp3 <- columnStore.initializeProjection(datasetObj.projections.head) }
+      yield {
+        originator ! DatasetCreated
+      }).recover {
+        case e: StorageEngineException => originator ! e
+      }
+    }
+  }
+
+  // If the coordinator is already set up, then everything is already fine.
+  // Otherwise get the dataset object and create a new actor, re-initializing state.
+  private def setupIngestion(originator: ActorRef,
+                             dataset: String,
+                             columns: Seq[String],
+                             version: Int,
+                             defaultPartKey: Option[PartitionKey]): Unit = {
+    def notify(msg: Any): Unit = { self ! DatasetCreateNotify(dataset, version, msg) }
+
+    if (dsCoordinators contains (dataset -> version)) {
+      originator ! IngestionReady
+    } else if (dsCoordNotify contains (dataset -> version)) {
+      // There is already a setupIngestion / dsCoordActor creation in progress.  Add to list of callbacks
+      // for the final result.
+      dsCoordNotify((dataset -> version)) = originator :: dsCoordNotify((dataset -> version))
+    } else {
+      dsCoordNotify((dataset -> version)) = List(originator)
+      // Everything after this point happens in a future, asynchronously from the actor processing.
+      // Thus 1) don't modify internal state, and 2) make sure we don't have multiple actor creations
+      // happening in parallel, thus the need for dsCoordNotify.
+      (for { datasetObj <- metaStore.getDataset(dataset)
+             schema <- verifySchema(originator, dataset, version, columns) if schema.isDefined }
+      yield {
+        val columnSeq = columns.map(schema.get(_))
+        // Create the RichProjection, and ferret out any errors
+        val proj = RichProjection.make(datasetObj, columnSeq)
+        proj.recover {
+          case RichProjection.BadSchema(reason) => notify(BadSchema(reason))
+          case Dataset.BadPartitionColumn(reason) => notify(BadSchema(reason))
+        }
+        proj.foreach { richProj =>
+          // Create the dataset coordinator
+          val typedProj = richProj.asInstanceOf[RichProjection[richProj.helper.Key]]
+          val props = DatasetCoordinatorActor.props(typedProj, version, columnStore, reprojector, config)
+          val ref = context.actorOf(props, s"ds-coord-${datasetObj.name}-$version")
+          self ! AddDatasetCoord(dataset, version, ref)
+          notify(IngestionReady)
+        }
+      }).recover {
+        case NotFoundError(what) => notify(UnknownDataset)
+        case t: Throwable        => notify(MetadataException(t))
+      }
+    }
+  }
```

Comment:
IMO this is big and a big procedural. I'd break it up to smaller functions with sub private functions.